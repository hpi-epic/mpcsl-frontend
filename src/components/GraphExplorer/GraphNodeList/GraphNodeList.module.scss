.NodeSelect {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 0.5rem;
  align-items: center;
  cursor: pointer;
  z-index: 1800;
  &:hover {
    background-color: rgba(0, 126, 255, 0.1);
  }
}

.NodeSelectDisabled {
  opacity: 0.5;
}
