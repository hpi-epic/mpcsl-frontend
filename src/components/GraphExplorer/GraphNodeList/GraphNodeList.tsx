import React, { useState, useEffect } from 'react';
import { Tooltip, Icon, Menu, Col, Popconfirm } from 'antd';
import { ID3GraphNode } from '../../../types/graphTypes';
import Select from 'react-virtualized-select';
import { isArray } from 'util';
import { GraphSingleton, GraphChanges } from '../../../graph/graph';
import { useParams } from 'react-router-dom';
import { filter } from 'rxjs/operators';

const GraphExplorerSelect = () => {
  const { resultId } = useParams<{ resultId: string }>();
  const [graphSearch, setGraphSearch] = useState<
    {
      value: number;
      label: string;
    }[]
  >([]);
  useEffect(() => {
    const sub = GraphSingleton.subject
      .pipe(
        filter(
          e =>
            e === GraphChanges.AvailableNodeChanged ||
            e === GraphChanges.NodesChanged
        )
      )
      .subscribe(() => {
        if (GraphSingleton.availableNodes) {
          const result = GraphSingleton.availableNodes
            .filter(
              aNode =>
                !GraphSingleton.nodes.some(
                  node => !node.isContext && node.id === aNode.id.toString()
                )
            )
            .map(node => ({
              value: node.id,
              label: node.name
            }));
          setGraphSearch(result);
        } else {
          setGraphSearch([]);
        }
      });
    return () => sub.unsubscribe();
  }, []);
  return (
    <div
      style={{
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-evenly'
      }}
    >
      <span
        style={{
          width: '70%'
        }}
      >
        <Select
          key="a"
          onChange={option =>
            option && !isArray(option) && option.value
              ? GraphSingleton.addNode(option.value, parseInt(resultId))
              : null
          }
          options={
            graphSearch
              ? (graphSearch.filter(n => !!n) as {
                  value: number;
                  label: string;
                }[])
              : []
          }
          onSelectResetsInput={true}
          onBlurResetsInput={false}
          valueKey="value"
          labelKey="label"
          closeOnSelect={false}
          removeSelected={true}
          clearable={true}
          placeholder="Select nodes"
        />
      </span>
      <Tooltip title="Load all nodes">
        <Popconfirm
          title="Are you sure that you want to load all nodes?"
          onConfirm={() => {
            GraphSingleton.addNodes(
              graphSearch.map(node => node.value),
              parseInt(resultId)
            );
          }}
        >
          <Icon
            style={{ color: 'rgba(255, 255, 255, 0.65)', fontSize: 20 }}
            type="plus-circle"
          />
        </Popconfirm>
      </Tooltip>
    </div>
  );
};

interface IPropsGraphNodeList {
  nodes: ID3GraphNode[];
  onNodeClick?: (node: ID3GraphNode) => void;
  onRemoveNode?: (nodeID: ID3GraphNode) => void;
  isSelectionMode: boolean;
}

const GraphNodeList = (props: IPropsGraphNodeList) => {
  return (
    <div
      style={{
        backgroundColor: '#001529',
        height: '100%'
      }}
    >
      {props.isSelectionMode ? <GraphExplorerSelect /> : null}
      <Menu theme="dark" selectable={false}>
        <Menu.ItemGroup title="Focused Nodes:">
          {props.nodes
            .filter(node => !node.isContext)
            .map(node => (
              <Menu.Item key={node.id}>
                <Col
                  span={18}
                  style={{ overflow: 'hidden' }}
                  onClick={() =>
                    props.onNodeClick ? props.onNodeClick(node) : undefined
                  }
                >
                  <Tooltip title={node.label}>{node.label}</Tooltip>
                </Col>
                {props.isSelectionMode ? (
                  <Col span={4} offset={2}>
                    <Icon
                      type="close"
                      onClick={() =>
                        props.onRemoveNode
                          ? props.onRemoveNode(node)
                          : undefined
                      }
                    />
                  </Col>
                ) : null}
              </Menu.Item>
            ))}
        </Menu.ItemGroup>
      </Menu>
    </div>
  );
};

export { GraphNodeList };
