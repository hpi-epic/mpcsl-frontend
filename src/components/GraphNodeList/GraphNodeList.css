.Node-List-Item {
  overflow: hidden;
  text-overflow: ellipsis;
}

.Node-List-Item > .ant-list-item-content {
  max-width: 200px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: block;
}
