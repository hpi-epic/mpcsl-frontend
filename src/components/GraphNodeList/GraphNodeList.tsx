import React, { useState, useEffect } from 'react';
import { Tooltip, Icon, Menu, Col, Popconfirm } from 'antd';
import { ID3GraphNode, IAPIGraphNode } from '../../types/graphTypes';
import { ThunkDispatch } from 'redux-thunk';
import { connect } from 'react-redux';
import * as actions from '../../actions/graphExplorer';
import { IState } from '../../store';
import Select from 'react-virtualized-select';
import { isArray } from 'util';

interface IPropsGraphNodeList {
  nodes: ID3GraphNode[];
  onNodeClick?: (node: ID3GraphNode) => void;
  onRemoveNode?: (nodeID: ID3GraphNode) => void;
  isSelectionMode: boolean;
}

interface IGraphExplorerProps {
  onAddNode: (nodeID: number) => void;
  availableNodes: IAPIGraphNode[];
  nodes: ID3GraphNode[];
}

const mapStateToProps = (state: IState) => {
  return {
    availableNodes: state.graphExplorer!.availableNodes,
    nodes: state.graphExplorer!.nodes
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<IState, void, actions.GraphExplorerAction>
) => {
  return {
    onAddNode: (nodeID: number) => dispatch(actions.addNode(nodeID))
  };
};

const GraphExplorerSelect = (props: IGraphExplorerProps) => {
  const [graphSearch, setGraphSearch] = useState<
    {
      value: number;
      label: string;
    }[]
  >([]);
  useEffect(() => {
    if (props.availableNodes) {
      const result = props.availableNodes
        .filter(node => !props.nodes.some(n => node.id.toString() === n.id))
        .map(node => ({
          value: node.id,
          label: node.name
        }));
      setGraphSearch(result);
    } else {
      setGraphSearch([]);
    }
  }, [props.availableNodes, props.nodes]);
  return (
    <div
      style={{
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-evenly'
      }}
    >
      <span
        style={{
          width: '70%'
        }}
      >
        <Select
          key="a"
          onChange={option =>
            option && !isArray(option) && option.value
              ? props.onAddNode(option.value)
              : null
          }
          options={
            graphSearch
              ? (graphSearch.filter(n => !!n) as {
                  value: number;
                  label: string;
                }[])
              : []
          }
          onSelectResetsInput={true}
          onBlurResetsInput={false}
          valueKey="value"
          labelKey="label"
          closeOnSelect={false}
          removeSelected={true}
          clearable={true}
          placeholder="Select nodes"
        />
      </span>
      <Tooltip title="Load all nodes">
        <Popconfirm
          title="Are you sure that you want to load all nodes?"
          onConfirm={() => {
            for (const node of graphSearch) {
              if (node) {
                props.onAddNode(node.value);
              }
            }
          }}
        >
          <Icon
            style={{ color: 'rgba(255, 255, 255, 0.65)', fontSize: 20 }}
            type="plus-circle"
          />
        </Popconfirm>
      </Tooltip>
    </div>
  );
};

const GraphExplorerSelectRedux = connect(
  mapStateToProps,
  mapDispatchToProps
)(GraphExplorerSelect);

function GraphNodeList(props: IPropsGraphNodeList) {
  return (
    <div
      style={{
        backgroundColor: '#001529',
        height: '100%'
      }}
    >
      {props.isSelectionMode ? <GraphExplorerSelectRedux /> : null}
      <Menu theme="dark" selectable={false}>
        <Menu.ItemGroup title="Focused Nodes:">
          {props.nodes.map(node => (
            <Menu.Item key={node.id}>
              <Col span={18} style={{ overflow: 'hidden' }}>
                <Tooltip title={node.label}>{node.label}</Tooltip>
              </Col>
              {props.isSelectionMode ? (
                <Col span={4} offset={2}>
                  <Icon
                    type="close"
                    onClick={() => props.onRemoveNode!(node)}
                  />
                </Col>
              ) : null}
            </Menu.Item>
          ))}
        </Menu.ItemGroup>
      </Menu>
    </div>
  );
}

export default GraphNodeList;
