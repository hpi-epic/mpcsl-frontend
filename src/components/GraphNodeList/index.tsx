import React from 'react';
import { List } from 'antd';

interface IPropsGraphNodeList {
  nodes: string[];
}

function GraphNodeList(props: IPropsGraphNodeList) {
  return (
    <List
      header={
        <span style={{ padding: '14px', fontWeight: 'bold' }}>
          Focused Nodes:
        </span>}
      size='small'
      dataSource={props.nodes}
      renderItem={(item: string) => (
        <List.Item style={{ padding: '14px' }}>{item}</List.Item>
      )}
    />
  );
}

export default GraphNodeList;
