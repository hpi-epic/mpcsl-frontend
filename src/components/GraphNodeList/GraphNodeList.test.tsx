import React from 'react';
import renderer from 'react-test-renderer';
import GraphNodeList from './GraphNodeList';

describe('<GraphNodeList />', () => {
  it('renders correctly', () => {
    const nodeList = [
      { id: '1', label: 'node1' },
      { id: '2', label: 'node2' },
      { id: '3', label: 'node3' }
    ];
    const tree = renderer
      .create(
        <GraphNodeList
          nodes={nodeList}
          onNodeClick={() => undefined}
          onRemoveNode={() => undefined}
          isSelectionMode={false}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
