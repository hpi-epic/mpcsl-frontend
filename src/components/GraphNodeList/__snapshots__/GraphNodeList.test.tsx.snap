// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GraphNodeList /> renders correctly 1`] = `
<div
  className="ant-list Node-List-Item ant-list-sm ant-list-split"
>
  <div
    className="ant-list-header"
  >
    <span
      style={
        Object {
          "fontWeight": "bold",
          "padding": "14px",
        }
      }
    >
      Focused Nodes:
    </span>
  </div>
  <div
    className="ant-spin-nested-loading"
    style={null}
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-list-item Node-List-Item"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "paddingBottom": "4px",
            "paddingLeft": "14px",
            "paddingTop": "4px",
          }
        }
      >
        <div
          className="ant-list-item-content ant-list-item-content-single"
        >
          <div
            onClick={[Function]}
          >
            node1
          </div>
        </div>
      </div>
      <div
        className="ant-list-item Node-List-Item"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "paddingBottom": "4px",
            "paddingLeft": "14px",
            "paddingTop": "4px",
          }
        }
      >
        <div
          className="ant-list-item-content ant-list-item-content-single"
        >
          <div
            onClick={[Function]}
          >
            node2
          </div>
        </div>
      </div>
      <div
        className="ant-list-item Node-List-Item"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "paddingBottom": "4px",
            "paddingLeft": "14px",
            "paddingTop": "4px",
          }
        }
      >
        <div
          className="ant-list-item-content ant-list-item-content-single"
        >
          <div
            onClick={[Function]}
          >
            node3
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
